ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala

```diff
@@ -403,31 +407,30 @@ class WebSocketService(
   private def emitOffsetTicksAndFilterOutEmptySteps[Pos](startFrom: Option[domain.StartingOffset])
     : Flow[StepAndErrors[Pos, JsValue], StepAndErrors[Pos, JsValue], NotUsed] = {
 
-    type TickTriggerOrStep = Unit \/ StepAndErrors[Pos, JsValue]
-
-    val tickTrigger: TickTriggerOrStep = -\/(())
-    val zeroState: StepAndErrors[Pos, JsValue] = startFrom.cata(
+    val zeroStep: StepAndErrors[Pos, JsValue] = startFrom.cata(
       x => StepAndErrors(Seq(), LiveBegin(AbsoluteBookmark(x.offset))),
       StepAndErrors(Seq(), LiveBegin(LedgerBegin))
     )
     Flow[StepAndErrors[Pos, JsValue]]
-      .map(a => \/-(a): TickTriggerOrStep)
-      .keepAlive(config.heartBeatPer, () => tickTrigger)
-      .scan((zeroState, tickTrigger)) {
-        case ((state, _), -\/(())) =>
-          // convert tick trigger into a tick message, get the last seen offset from the state
+      .map(a => Step(a))
+      .keepAlive(config.heartBeatPer, () => TickTrigger)
+      .scan((zeroStep, TickTrigger: TickTriggerOrStep[Pos])) {
+        case ((_, TickTrigger), TickTrigger) =>
+          // skip all TickTriggers preceding the initial ACS retrieval
+          (zeroStep, TickTrigger)
+        case ((state, _), TickTrigger) =>
```

Comment:
Could we just replace `state` by `Option[domain.StartingOffset]`? I can’t find any case where we use more info than is provided by this and it seems to make the code much clearer: None means you haven’t seen the offset, `Some` means you have seen it (we still need the second element of the pair).

Reply:
good idea! Thanks!

Reply:
looks like it is a bit harder to make that change than I originally though, too many offsets, of different types.

Reply:
but I think it reads better now.